### PR TITLE
Avoid unnecessary flattening in all-gather and reduce-scatter

### PR DESCRIPTION
**Pull request by thorjohnsen**
This PR addresses an important special case, which comes up frequently as we are working towards model parallel training and a few other things. torch.distributed.all_gather takes two arguments, a list of output tensors and a single input tensor. The all_gather op collects inputs from all the ranks into the list of output tensors. Internally, the rank inputs are collected into a flattened tensor and then copied from there to the output tensors. Often, the list of output tensors are simply views into an already flattened tensor, in which case the un-flattening is unnecessary and we can do an in-place all-gather op instead. This saves memory and improves performance. This is particularly important when the all-gather is done within a single node, because all the reductions are done over high-speed nvlinks and the extra D2D copies really hurt performance. To demonstrate the difference this PR makes, I ran the test script shown below on a DGX1 with 8 x 32GB V100 cards, using NVIDIA's 19.11 devel image. Without this PR, it took 13.96 seconds (average of 5 runs) to complete 100 all-gathers of a flattened tensor with 1.07 billion floats (1024^3). With this PR, the same 100 all-gathers took 3.01 seconds, a 4.64x improvement.

The situation for reduce-scatter is fundamentally the same, but reversed. Instead of doing D2D copies after the op to un-flatten the output tensors, reduce-scatter has to do D2D copies before the op to flatten the input tensors. The performance overhead is the same as for all-gather. This PR fixes both all-gather and reduce-scatter.

This PR should have no side-effects. Everything should work exactly like before, but when the special case is detected, you will see a large bump in throughput.

Thanks to @alpha0422 for contributing the original code. 

Command line:
`python -m torch.distributed.launch --nnodes=1 --nproc_per_node=8 test_all_gather.py`

test_all_gather.py
```
import time
import torch

def main(args):
    a = torch.randn([args.group_size*args.size]).float().cuda()
    a_vec = [a[i*args.size:(i+1)*args.size] for i in range(args.group_size)]
    if torch.distributed.get_rank() == 0:
        print("We used up %d mb of GPU memory." % (torch.cuda.memory_allocated()/(1024*1024)))
        print("Doing %d all-gathers of %d x %d floats" % (args.niter, args.group_size, args.size))
    torch.distributed.barrier()
    before = time.time()
    for iter in range(args.niter):
        torch.distributed.all_gather(a_vec,a_vec[torch.distributed.get_rank()])
    torch.cuda.synchronize()
    torch.distributed.barrier()
    after = time.time()
    print("rank %d :: a.norm() = %e, run_time was %.2f seconds" % (torch.distributed.get_rank(), a.norm(), after-before))

if __name__ == '__main__':
    torch.distributed.init_process_group("nccl")
    torch.cuda.set_device(torch.distributed.get_rank() % torch.cuda.device_count())
    torch.manual_seed(888)
    class Args:
        size = 1024*1024*128
        niter = 100
        group_size = torch.distributed.get_world_size()
    main(Args())
```

### CODE REVIEW

**Comment by mrshenli** [comment]
cc @agolynski Does this PR has implications on all_gather_coalesced and all_gather_base/single?

**Comment by agolynski** [comment]
> cc @agolynski Does this PR has implications on all_gather_coalesced and all_gather_base/single?

This is a useful PR (at least in the current state of code).

ASIS, this does not seem to play well with coalesced impls as tensors are assumed to be of the same size, e.g.
t.storage_offset() != (tensor_lists[i][0].storage_offset() +
                    j * other[i].numel())

but i don't see a reason why this can't be extended to work with coalesced tensors as well.

The eventual direction is to have 
all_gather_base which will guarantee that the outputs are 'inplace' instead of discovering it in code as in this PR. However, If clients call into PG all_gather with custom allocated output tensors, this optimization is still useful.
Does it make sense?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -332,7 +332,8 @@ void check_gpu_tensors(const std::vector<at::Tensor>& tensors) {
 std::vector<at::Tensor> flatten_for_scatter_gather(
     std::vector<std::vector<at::Tensor>>& tensor_lists,
     std::vector<at::Tensor>& other,
-    size_t world_size) {
+    size_t world_size,
+    size_t rank) {
```
Comment (agolynski) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: to simplify reasoning in upstream code, can you introduce 
bool* inplace or bool* already_flat instead and fill it in as you do below for 'inplace'?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -367,8 +368,32 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
         );
       }
     }
-    // Flatten the tensors (from all ranks) into a single big tensor.
-    flattened[i] = newLikeFlat(tensor_lists, i);
+
+    //  Check if the tensors are already flattened to do inplace operation
+    bool inplace = true;
+    for (auto j = size_t{}; j < tensor_lists[i].size(); ++j) {
+        auto t = tensor_lists[i][j];
```
Comment (agolynski) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: nit: i think we have 2 space indents

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -367,8 +368,32 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
...
+        if (!t.storage().is_alias_of(other[i].storage()) ||
```
Comment (agolynski) [inline torch/lib/c10d/ProcessGroupNCCL.cpp]: I might be confused about this. Why outputs should share storage with inputs? Don't we just want all outputs to share storage or this is a special case of allocating storage where outputs are combined with inputs (I didn't know this is allowed by NCCL engine!). 
If that's the case, this saves some memory for separate input allocation in addition to output allocation!

However, for this specific case, maybe it's just enough to check if all outputs are properly aligned? 
Do you know what would happen if outputs are aligned and share storage with inputs but inputs aren't properly placed within outputs, would NCCL crash or produce garbage outputs?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (thorjohnsen) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Do you mean introduce a new flag in python API so that "discovery" logic can be removed from ProcessGroupNCCL.cpp? In other words

torch.distributed.all_gather(output_tensors, input_tensor)

Will always flatten output tensors, while

torch.distributed.all_gather(output_tensors, input_tensor, already_flat=True)

will always do an inplace reduction?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (thorjohnsen) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Will fix

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (thorjohnsen) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: NCCL library allows inplace operations, it's explicitly mentioned in the documention. The inputs have to be properly aligned. I don't know what happens if they are not, other than it will not work

image

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (agolynski) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: This can be an option as well, but I think the point of this cl is to allow user code to run fast without necessary changing it explicitly, so they'll get it 'for free'.

I meant having output bool here, e.g.
auto outputFlattened = flatten_for_scatter_gather(
    outputTensors, inputTensors, size, rank, *output_is_flat
  );

and then use it instead of

 if (outputFlattened[i][j].storage().is_alias_of(
              outputTensors[i][j].storage()) &&
              outputTensors[i][j].storage_offset() ==
              (outputTensors[i][0].storage_offset() +
              outputTensors[i][j].numel() * j)) {
             break;
          }

Would it work?

**Comment by agolynski** [comment]
Would it be possible to add a test that exercises the new code path?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (agolynski) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: That makes sense. In-place will definitely save memory in addition to copying data. 

I was wondering if we can save copying in case we've allocated output tensors separately from inputs, but outputs are still contiguous (that is, already flat). Something like:

input = torch.tensor(size)
....
outputs = torch.tensor(size*world_size*num_devices)
pg.allgather(output, input)

Do you think this usecase is worth covering as well?

(Alternatively we can just ask users to copy 'input' at the appropriate location inside 'output' and change its storage accordingly.)

What do you think?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (thorjohnsen) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: It will work, but the logic that determines if the list of tensors are views into an already flattened tensor will then have to go into the code that implements each NCCL op, so we'd end up duplicating it. Unless I put the reasoning code in a separate method.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (thorjohnsen) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Both cases should work. For all-gather, we can avoid copying to flatten and un-flatten if the outputs are views into an already flattened tensor. In addition, if the input is also a view into the same flattened tensor and input is properly aligned, we can do an in-place operation. The code in my PR already does this, but I don't verify that the input is correctly aligned. I will add code to verify proper alignment.

**Comment by thorjohnsen** [comment]
Absolutely! I can add a test script to verify correctness and demonstrate the perf benefit.

**Comment by thorjohnsen** [comment]
@mrshenli mentioned all_gather_coalesce as a potential issue. Is this something I need to address in this PR?

**Comment by agolynski** [comment]
I've discussed this with @mrshenli offline and he mentioned that this might become a breaking change if the following holds:
* operation is performed async
* tensors are positioned for in-place operation
* client changes input tensor as operation is in-flight

previous behavior: we make a copy and correct copy is being distributed
new behavior: undefined as tensor is being modified as op is in-flight

the suggestion is to have this in API explicitly, e.g. add optional bool in_place = false to python, and similarly for C++. You'll still get your performance gains by setting in_place = true, but we can avoid breaking clients. 

What do you think?

**Comment by thorjohnsen** [comment]
I think that is a sensible compromise. I will make the change.

**Comment by thorjohnsen** [comment]
I have added inplace flag to torch.distributed.all_reduce and torch.distributed.all_gather. Please have a look and let me know if this is not what you had in mind. I am working on a couple of test scripts and will submit them when they are ready.

**Comment by agolynski** [comment]
> I have added inplace flag to torch.distributed.all_reduce and torch.distributed.all_gather. Please have a look and let me know if this is not what you had in mind. I am working on a couple of test scripts and will submit them when they are ready.

Looks good, thank you!

**Comment by ddkalamk** [comment]
> * client changes input tensor as operation is in-flight

@agolynski Is this even allowed today from API point of view? As MPI backend simply enqueues the operation to MPI thread and returns. It may produce wrong results if input tensors are modified before `wait` is called and completed, I am missing something?

**Comment by ddkalamk** [comment]
> the suggestion is to have this in API explicitly, e.g. add optional bool in_place = false to python, and similarly for C++. You'll still get your performance gains by setting in_place = true, but we can avoid breaking clients.

IMHO, it is better to do in-place check internally as user can easily make a mistake here. And check itself would be very simple and inexpensive in most cases.

**Comment by thorjohnsen** [comment]
Hi @ddkalamk the inplace flag simply signals that NCCL inplace op is allowed. As you have pointed out, inplace NCCL op can potentially fail if user modifies input tensors while op is in flight, hence this is an opt-in feature, i.e. the inplace flag defaults to False. The code still checks internally if arguments allow inplace op and will only schedule an inplace NCCL op if inplace == True and the arguments support it.

**Comment by ddkalamk** [comment]
> inplace NCCL op can potentially fail if user modifies input tensors while op is in flight, hence this is an opt-in feature

Modifying inputs while op is in flight is always disallowed whether we are doing in-place or out-of-place operation as we don't get to know when op has finished reading the input. Even MPI backend would potentially fail if one modifies input before op is complete. So, not sure how this argument helps. On the other hand, what happens if input and output buffers overlap and user doesn't specify `inplace=True`? Or the other way?

I am not familiar with NCCL but from API description it appears that if input and output pointers are overlapping and correctly aligned, it would implicitly do in-place operation. But least in MPI case, one need to explicitly specify `MPI_IN_PLACE` for send buf when using in-place operation.

**Comment by ddkalamk** [comment]
> previous behavior: we make a copy and correct copy is being distributed

Sorry, i think I misunderstood this comment. @agolynski do you mean if tensors are place for in-place operation (i.e. overlapping), we make a copy of input and to make it explicitly out-of-place operation? And to disable this behavior we need extra argument? Is this behavior specific to NCCL or applies to other backends as well?

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -1131,6 +1135,12 @@ def all_gather(tensor_list,
         tensor (Tensor): Tensor to be broadcast from current process.
         group (ProcessGroup, optional): The process group to work on
         async_op (bool, optional): Whether this op should be an async op
+        no_copy (bool, optional): Do not unflatten output tensors if they
+            are already contiguous views into an already flattened tensor
+            and backend is NCCL. If input tensor shares storage with
+            output tensors, it must be properly aligned, i.e.
+            offset == rank * flat_output.numel() // world_size.
```
Comment (agolynski) [inline torch/distributed/distributed_c10d.py:1142]: Could you be a bit more specific here, e.g. something like
"""
all input and output tensors must be in the same storage and properly aligned, i.e. offset(input tensor in i-th rank) = i * num_elements(input tensor)
"""

**Comment by agolynski** [comment]
sorry for delays with reviewing this, @thorjohnsen. let me know when you get a chance to add testing for this

**Comment by thorjohnsen** [comment]
Hi @agolynski I have added the following tests to test/distributed/test_c10d.py

test_reduce_scatter_no_copy_ops -> same as test_reduce_scatter_ops, except inputs are flattened before call to avoid flattening of inputs

test_reduce_scatter_inplace_ops -> same as test_reduce_scatter_no_copy_ops, except inputs and outputs share storage

test_allgather_no_copy_ops -> same as test_allgather_ops, except outputs are flattened before call to avoid unflattening of outputs

test_allgather_inplace_ops -> same as test_allgather_no_copy_ops, except inputs and outputs share storage

File: torch/distributed/distributed_c10d.py (hunk 4 above)
Comment (thorjohnsen) [reply torch/distributed/distributed_c10d.py:1170]: Hi, there are actually two modes here. The performance benefit we are after here is gained by not having to unflatten the outputs if the output tensors are views into an already flattened tensor. The input tensor does not have to share storage with the outputs, but this is allowed and can save some memory. If input and outputs does share storage, input has to be properly aligned, that's what I was trying to convey with that paragraph. I'll try to rephrase it to make it clearer.

**Comment by csarofeen** [comment]
@agolynski is there anything else major that needs to be done for this PR or if @thorjohnsen fixes the flake8 errors will you attempt to land this?

**Comment by agolynski** [comment]
> @agolynski is there anything else major that needs to be done for this PR or if @thorjohnsen fixes the flake8 errors will you attempt to land this?

@thorjohnsen thanks for the tests! taking another look now.

**Comment by thorjohnsen** [comment]
How do I proceed with this PR? My last 3 commits only changed whitespace or wording inside comments, before that all tests were passing except flake8. Flake8 is now passing, but some other tests are failing for reasons that seem unrelated to the code in this PR. For instance, caffe2_onnx_main_py3_6_clang7_ubuntu16_04_build fails with the following error message:

Apr 27 16:58:18 Unable to checkout '815e209e4f30910dfdc47aa2ad041e3a46d61b44' in submodule path 'third_party/fbgemm'

815e209e4f30910dfdc47aa2ad041e3a46d61b44 is not one of my commits. Do I simply wait until these issues are resolved? I need to give progress report to management. Grateful for any help!

**Comment by pritamdamania87** [comment]
@thorjohnsen PR looks good to me, could you rebase and resubmit and I'm happy to merge in it. @agolynski Let me know if you have any further comments on this PR.

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 5]
```diff
@@ -610,8 +612,37 @@ std::vector<at::Tensor> flatten_for_scatter_gather(
...
+
+    if (no_copy) {
+      // no_copy operation requires all tensors in tensor_list are contiguous views
+      // into a single flattened tensor
+      for (auto j = size_t{}; j < tensor_lists[i].size(); ++j) {
+        auto t = tensor_lists[i][j];
+        if (!tensor_lists[i][j].storage().is_alias_of(tensor_lists[i][0].storage()) ||
+            tensor_lists[i][j].storage_offset() != (tensor_lists[i][0].storage_offset() +
+              j * tensor_lists[i][0].numel())) {
+          no_copy = false;
```
Comment (rohan-varma) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:624]: I wonder if it would be better to be explicit here and throw or at least warn if user specified no_copy but we can't actually do it?

**Comment by osalpekar** [comment]
Hi @thorjohnsen! Are you planning on continue work on this PR? I think we should be good to go after a rebase, and there are some important perf issues this can solve.

Do let us know!

**Comment by thorjohnsen** [comment]
Hi @osalpekar I'll try to get this rebased right now.

**Comment by osalpekar** [comment]
@thorjohnsen Thanks! Please let us know if you need any help.

**Comment by osalpekar** [comment]
@thorjohnsen Let us know if you can rebase by this Friday. Since merging this is a bit urgent, I'll have to take over the PR on Friday

**Comment by thorjohnsen** [comment]
@osalpekar If you have the time, I would be grateful if you can take over and bring this in. I am under a tight deadline and am unable to look at this PR until next week.

**Comment by slayton58** [comment]
@osalpekar I've taken a stab at rebasing here: https://github.com/slayton58/pytorch/tree/avoid_unnecessary_flattening but I've been unable to get tests to even run (they either abort or hang on my system, no output) -- it might help you some.

**Comment by zhaojuanmao** [comment]
Taking it over

**Comment by zhaojuanmao** [comment]
@thorjohnsen I can not import your PR, would you please help signing the CLA? or do you have time to rebase this PR? thanks.

**Comment by zhaojuanmao** [comment]
someone else is working on implementing all-gather-base that will just return a single flatten tensor, I think that implementation is more cleaner for use cases mentioned here, and can replace improvement in this PR. 

thoughts? @agolynski
